### PR TITLE
Add Backup and Restore hook's guides

### PR DESCRIPTION
**Pull request by hossainemruz**
Tasks:
- [x] Overview
- [x] Backup and Restore hooks
- [x] Batch Backup hooks
- [x] Configuring different types of hook
- [x] Update concept docs

### CODE REVIEW

File: docs/guides/latest/hooks/overview.md [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+- **Pre-Task Hook Failed:** If a pre-task hook fails to execute, the rest of the backup/restore process will be skipped and the respective `BackupSession`/`RestoreSession` will be marked as `Failed`. You may see following the things to happen in addition to skipping the backup process:
+  - **Backup Sidecar:** If the pre-task hook fails in the backup sidecar, the sidecar will just log the failure and continue watching for `BackupSession` for the next backup.
+  - **Restore Init-Container:** If the pre-task hook fails in restore init-container, the container will crash. Hence, your workload will be stuck in the initialization phase.
+  - **Backup or Restore Job:** If the pre-task hook fails in backup or restore job, the container will fail. Hence, the job will never go to the completed stage. You may see the job to create multiple pods to retry.
+
+- **Backup/Restore Process Failed:** The post-task hook will be executed even if the backup/restore process failed. This is to cover the scenario where you have paused your application in a pre-task hook and want to resume it in a post-task hook.
```
Comment (tamalsaha) [inline docs/guides/latest/hooks/overview.md]: `You may see the job to create multiple pods to retry.`

Should we allow retry? We may disable retry and fail the Job. I think that sounds better?

File: docs/guides/latest/hooks/configuring-hooks.md [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+    httpHeaders:
+    - name: "User-Agent"
+      value: "stash/0.9.x"
+```
+
+## HTTPPost
+
+`httpPost` hook allows sending an HTTP POST request to an HTTP server before and after the backup/restore process. The following configurable fields are available for `httpPost` hook.
+
+| Field         | Type       | Usage                                                                                                                                                                                                                                                                             |
```
Comment (tamalsaha) [inline docs/guides/latest/hooks/configuring-hooks.md]: We don't mention the data types in these fields. I think that might be a problem.

File: docs/guides/latest/hooks/backup-and-restore-hooks.md [hunk 3]
```diff
@@ -0,0 +1,720 @@
...
+    identifier: backup-and-restore-hooks
+    name: Backup & Restore Hooks
+    parent: hooks
+    weight: 20
+product_name: stash
+menu_name: docs_{{ .version }}
+section_menu_id: guides
+---
+
+# Backup & Restore Hooks
```
Comment (tamalsaha) [inline docs/guides/latest/hooks/backup-and-restore-hooks.md]: Should we write, "This is an advacned concept. Try simpler things first."

File: docs/guides/latest/hooks/backup-and-restore-hooks.md [hunk 4]
```diff
@@ -0,0 +1,720 @@
...
+
+```console
+$ kubectl get cronjob -n demo
+NAME                            SCHEDULE      SUSPEND   ACTIVE   LAST SCHEDULE   AGE
+stash-backup-backup-hook-demo   */5 * * * *   True      0        5m13s           29m
+```
+
+**Simulate Disaster Scenario:**
+
+Now, let's simulate a disaster scenario. Here, we are going to delete all the data from the `employee` table.
```
Comment (tamalsaha) [inline docs/guides/latest/hooks/backup-and-restore-hooks.md]: `DROP TABLE` might be even better.

File: docs/guides/latest/hooks/batch-backup-hooks.md [hunk 5]
```diff
@@ -0,0 +1,457 @@
...
+gcs-repo   true               2                72s                      18m
+```
+
+Here, `SNAPSHOT-COUNT` 2 indicates that one snapshot for each target has been taken successfully.
+
+### Verify Backup Hooks Executed
+
+Now, go to your slack channel. You should see that Stash has sent notification before and after the backup of WordPress application.
+
+<figure align="center">
```
Comment (tamalsaha) [inline docs/guides/latest/hooks/batch-backup-hooks.md]: Beautiful!

File: docs/concepts/crds/restoresession.md [hunk 6]
```diff
@@ -74,7 +89,7 @@ spec:
     pod:
       imagePullSecrets:
       - name: my-private-registry-secret
-      serviceAccountName: my-restore-svc
+      serviceAccountName: my-backup-svc
```
[the remark this answers is not in the document]
Comment (hossainemruz) [reply docs/concepts/crds/restoresession.md]: Shouldn't it be `restore`? It's `RestoreSession` spec. He is trying to restore after disaster. So, how the `backup` sa will exist then?

File: docs/guides/latest/hooks/overview.md (hunk 1 above)
Comment (hossainemruz) [reply docs/guides/latest/hooks/overview.md]: We already set `backOffLimit` to 1. But, it doesn't work. :(
https://github.com/stashed/stash/blob/ba1733ebc9bcc86524613c75ef82bc2a6b729c73/pkg/controller/backup_session.go#L312

File: docs/guides/latest/hooks/configuring-hooks.md (hunk 2 above)
Comment (hossainemruz) [reply docs/guides/latest/hooks/configuring-hooks.md]: I wanted to mention data type. But, we can't do that when there is nested type.

File: docs/guides/latest/hooks/backup-and-restore-hooks.md (hunk 3 above)
Comment (hossainemruz) [reply docs/guides/latest/hooks/backup-and-restore-hooks.md]: This doc already comes at the bottom of the sidebar. Isn't it natural to try the upper docs first? Btw, I agree to include.

File: docs/guides/latest/hooks/configuring-hooks.md (hunk 2 above)
Comment (hossainemruz) [reply docs/guides/latest/hooks/configuring-hooks.md]: done

File: docs/concepts/crds/restoresession.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (tamalsaha) [reply docs/concepts/crds/restoresession.md]: I agree. This should stay `restore`.

File: docs/guides/latest/hooks/overview.md (hunk 1 above)
Comment (tamalsaha) [reply docs/guides/latest/hooks/overview.md]: https://kubernetes.io/docs/concepts/workloads/controllers/jobs-run-to-completion/#handling-pod-and-container-failures

`Note that even if you specify .spec.parallelism = 1 and .spec.completions = 1 and .spec.template.spec.restartPolicy = "Never", the same program may sometimes be started twice.`

I guess there is nothing we can do.

**Pull request merged**